keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,313 @@
+---
+
+# Immutable fields
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+A lot of fields in APIs tend to be "immutable", they can't be changed after
+creation. This is true for example for many of the fields in pods. There is
+currently no way to declaratively specify that fields are immutable, and one
+has to rely on either built-in validation for core types, or have to build a
+validating webhooks for CRDs.
+
+Providing a new `// +immutable` marker would help 
+- to make the API more descriptive to users
+- to help API developers by verifying these assertions automatically
+- and to publish this information via OpenAPI.
+
+## Motivation
+
+There are resources in Kubernetes which have immutable fields by design,
+i.e. after creation of an object, those fields cannot be mutated anymore. E.g. a
+pod's specification is mostly unchangeable once it is created. To change the
+pod, it must be deleted, recreated and rescheduled. Users want to implement the
+same kind of read-only semantics for CustomResources, for example:
+https://github.com/kubernetes/kubernetes/issues/65973. Today this is only possible
+with the unreasonable development overhead of a webhook.
+
+### Goals
+
+- extend the CRD API to be able to specify immutability for fields.
+- publish the immutability field of CRDs via OpenAPI as vendor extension.
+- verify immutability on CR update and patch requests.
+- propose a source code marker to be consumed by kubebuilder and openapi-gen.
+- the semantics of immutability must be driven by:
+  - we do not break/change old CRD persistence semantics.
+  - the user-observed equality used for immutability checks must match the equality on
+    persisted objects. I.e. if `StorageRoundtrip(object)` is the object returned by a
+    create or update call, then we want that `StorageRoundtrip(a) == StorageRoundtrip(b)`
+    is the equality used for comparing `a` and `b`. If that check fails, a request
+    is rejected because of immutability conflict.
+- the mechanism must extend to
+  - the addition of protobuf or other encodings which unify values like empty, null and undefined.
+  - the use for existing native types in order to replace complex validation code with a simple declarative marker on the types.
```

Comment:
@erictune mentioned a requirement that this proposal supports with but might be worth calling in the goals here: create requests that set immutable fields must be idempotent. It highlights the importance of handling equality correctly.

Reply:
do you have an example?

Reply:
Here's the original comment: https://github.com/kubernetes/enhancements/pull/1099#discussion_r296360894

It's got an example in it.